[cond] inlining into one of the branches when pred is a python constant

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128709

When the input predicate is a python constant, we specialize into one of the branches and warn users that torch.cond is not preserving the dynamism. The previous behavior is that we baked in True/False in the cond operator. This can be confusing. In this PR, we change it to be specializing into one of the branches when the inputs are constants.

We additionally change the naming of cond operator to default one without overriding its name. This allows better testing on de-serialized graph.

Test Plan:
The predicate in some existing tests is the result of a shape comparison. When no dynamic shape is involved, the predicate is a python bool. To fix them, we either change the predicate to be some data-dependent tensor or change the test to check cond is specialized as one of the branches, 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

Differential Revision: [D59589709](https://our.internmc.facebook.com/intern/diff/D59589709)

ydwu4 committed:
[cond] inlining into one of the branches when pred is a python constant

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] inlining into one of the branches when pred is a python constant"


When the input predicate is a python constant, we specialize into one of the branches and warn users that torch.cond is not preserving the dynamism.

The predicate in some existing tests is the result of a shape comparison. When no dynamic shape is involved, the predicate is a python bool. To fix them, we either change the predicate to be some data-dependent tensor or change the test to check cond is specialized as one of the branches, 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] inlining into one of the branches when pred is a python constant"


When the input predicate is a python constant, we specialize into one of the branches and warn users that torch.cond is not preserving the dynamism.

The predicate in some existing tests is the result of a shape comparison. When no dynamic shape is involved, the predicate is a python bool. To fix them, we either change the predicate to be some data-dependent tensor or change the test to check cond is specialized as one of the branches, 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] inlining into one of the branches when pred is a python constant"


When the input predicate is a python constant, we specialize into one of the branches and warn users that torch.cond is not preserving the dynamism.

The predicate in some existing tests is the result of a shape comparison. When no dynamic shape is involved, the predicate is a python bool. To fix them, we either change the predicate to be some data-dependent tensor or change the test to check cond is specialized as one of the branches, 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] inlining into one of the branches when pred is a python constant"


When the input predicate is a python constant, we specialize into one of the branches and warn users that torch.cond is not preserving the dynamism.

The predicate in some existing tests is the result of a shape comparison. When no dynamic shape is involved, the predicate is a python bool. To fix them, we either change the predicate to be some data-dependent tensor or change the test to check cond is specialized as one of the branches, 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

angelayi commented on torch/_higher_order_ops/cond.py:
```diff
@@ -107,10 +111,19 @@ def false_fn(x: torch.Tensor):
+    if isinstance(pred, (bool, int, float)):
+        log.warning(
+            "Pred is a Python constant. When used with torch.cond, it executes only one of the branches."
+            " If you want torch.cond to perserve two branches, please make the predicate a boolean tensor."
```
can cond take symbool?

ydwu4 replied:
yeah, add a test_export_cond_symbool_pred to reflect this. cond won't specialize on symbool.

ydwu4 committed:
Update on "[cond] inlining into one of the branches when pred is a python constant"


When the input predicate is a python constant, we specialize into one of the branches and warn users that torch.cond is not preserving the dynamism.

The predicate in some existing tests is the result of a shape comparison. When no dynamic shape is involved, the predicate is a python bool. To fix them, we either change the predicate to be some data-dependent tensor or change the test to check cond is specialized as one of the branches, 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] inlining into one of the branches when pred is a python constant"


When the input predicate is a python constant, we specialize into one of the branches and warn users that torch.cond is not preserving the dynamism.

The predicate in some existing tests is the result of a shape comparison. When no dynamic shape is involved, the predicate is a python bool. To fix them, we either change the predicate to be some data-dependent tensor or change the test to check cond is specialized as one of the branches, 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] inlining into one of the branches when pred is a python constant"


When the input predicate is a python constant, we specialize into one of the branches and warn users that torch.cond is not preserving the dynamism. The previous behavior is that we baked in True/False in the cond operator. This can be confusing. In this PR, we change it to be specializing into one of the branches when the inputs are constants.

Test Plan:
The predicate in some existing tests is the result of a shape comparison. When no dynamic shape is involved, the predicate is a python bool. To fix them, we either change the predicate to be some data-dependent tensor or change the test to check cond is specialized as one of the branches, 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 committed:
Update on "[cond] inlining into one of the branches when pred is a python constant"


When the input predicate is a python constant, we specialize into one of the branches and warn users that torch.cond is not preserving the dynamism. The previous behavior is that we baked in True/False in the cond operator. This can be confusing. In this PR, we change it to be specializing into one of the branches when the inputs are constants.

Test Plan:
The predicate in some existing tests is the result of a shape comparison. When no dynamic shape is involved, the predicate is a python bool. To fix them, we either change the predicate to be some data-dependent tensor or change the test to check cond is specialized as one of the branches, 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot merge

Cancelled old merge job to fix something, reissuing merge command, sorry about that

ydwu4 committed:
Update on "[cond] inlining into one of the branches when pred is a python constant"


When the input predicate is a python constant, we specialize into one of the branches and warn users that torch.cond is not preserving the dynamism. The previous behavior is that we baked in True/False in the cond operator. This can be confusing. In this PR, we change it to be specializing into one of the branches when the inputs are constants.

Test Plan:
The predicate in some existing tests is the result of a shape comparison. When no dynamic shape is involved, the predicate is a python bool. To fix them, we either change the predicate to be some data-dependent tensor or change the test to check cond is specialized as one of the branches, 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] inlining into one of the branches when pred is a python constant"


When the input predicate is a python constant, we specialize into one of the branches and warn users that torch.cond is not preserving the dynamism. The previous behavior is that we baked in True/False in the cond operator. This can be confusing. In this PR, we change it to be specializing into one of the branches when the inputs are constants.

We additionally change the naming of cond operator to default one without overriding its name. This allows better testing on de-serialized graph.

Test Plan:
The predicate in some existing tests is the result of a shape comparison. When no dynamic shape is involved, the predicate is a python bool. To fix them, we either change the predicate to be some data-dependent tensor or change the test to check cond is specialized as one of the branches, 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 commented:
@pytorchbot merge -f "failures not related to this pr"

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but it is breaking one ExecuTorch test' -c ghfirst

The failed test is from https://github.com/pytorch/executorch/blob/main/exir/tests/test_passes.py

```
buck2 test '@fbcode//mode/dev' fbcode//executorch/exir/tests:passes -- --exact 'executorch/exir/tests:passes - test_dce_recursive (executorch.exir.tests.test_passes.TestPasses)'

======================================================================
FAIL: test_dce_recursive (executorch.exir.tests.test_passes.TestPasses)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/5ce5662abd58612b/executorch/exir/tests/__passes__/passes#link-tree/executorch/exir/tests/test_passes.py", line 756, in test_dce_recursive
    self.assertTrue(torch.ops.aten.sub.Tensor in collect_ops(gm))
AssertionError: False is not true

----------------------------------------------------------------------
Ran 1 test in 0.362s
```

We are running ExecuTorch tests on PyTorch CI and should have caught this issue, but the OSS coverage is not there yet and the test is currently skipped on OSS CI https://github.com/pytorch/executorch/blob/main/pytest.ini#L51

Maybe someone from ExecuTorch could give a timeline on when this is enabled cc @guangy10 @kirklandsign

ydwu4 commented:
Try a re-merge after the fix in D59287189 is merged.

ydwu4 commented:
@pytorchbot merge

ydwu4 committed:
Update on "[cond] inlining into one of the branches when pred is a python constant"


When the input predicate is a python constant, we specialize into one of the branches and warn users that torch.cond is not preserving the dynamism. The previous behavior is that we baked in True/False in the cond operator. This can be confusing. In this PR, we change it to be specializing into one of the branches when the inputs are constants.

We additionally change the naming of cond operator to default one without overriding its name. This allows better testing on de-serialized graph.

Test Plan:
The predicate in some existing tests is the result of a shape comparison. When no dynamic shape is involved, the predicate is a python bool. To fix them, we either change the predicate to be some data-dependent tensor or change the test to check cond is specialized as one of the branches, 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 commented:
@ydwu4 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59589709).

ydwu4 commented:
@pytorchbot merge

ydwu4 commented:
@pytorchbot revert -m "causing error on truck due to a land racing: https://hud.pytorch.org/pytorch/pytorch/commit/fe3e6878c4bb2a6001045c179fd7fa9838242558"

ydwu4 commented:
@pytorchbot revert -m "causing error on truck due to a land racing: https://hud.pytorch.org/pytorch/pytorch/commit/fe3e6878c4bb2a6001045c179fd7fa9838242558" -c landrace

ydwu4 committed:
Update on "[cond] inlining into one of the branches when pred is a python constant"


When the input predicate is a python constant, we specialize into one of the branches and warn users that torch.cond is not preserving the dynamism. The previous behavior is that we baked in True/False in the cond operator. This can be confusing. In this PR, we change it to be specializing into one of the branches when the inputs are constants.

We additionally change the naming of cond operator to default one without overriding its name. This allows better testing on de-serialized graph.

Test Plan:
The predicate in some existing tests is the result of a shape comparison. When no dynamic shape is involved, the predicate is a python bool. To fix them, we either change the predicate to be some data-dependent tensor or change the test to check cond is specialized as one of the branches, 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D59589709](https://our.internmc.facebook.com/intern/diff/D59589709)

[ghstack-poisoned]

ydwu4 commented:
@ydwu4 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59589709).

ydwu4 commented:
@pytorchbot merge

ydwu4 committed:
Update on "[cond] inlining into one of the branches when pred is a python constant"


When the input predicate is a python constant, we specialize into one of the branches and warn users that torch.cond is not preserving the dynamism. The previous behavior is that we baked in True/False in the cond operator. This can be confusing. In this PR, we change it to be specializing into one of the branches when the inputs are constants.

We additionally change the naming of cond operator to default one without overriding its name. This allows better testing on de-serialized graph.

Test Plan:
The predicate in some existing tests is the result of a shape comparison. When no dynamic shape is involved, the predicate is a python bool. To fix them, we either change the predicate to be some data-dependent tensor or change the test to check cond is specialized as one of the branches, 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D59589709](https://our.internmc.facebook.com/intern/diff/D59589709)

[ghstack-poisoned]

ydwu4 commented:
@ydwu4 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59589709).

ydwu4 committed:
Update on "[cond] inlining into one of the branches when pred is a python constant"


When the input predicate is a python constant, we specialize into one of the branches and warn users that torch.cond is not preserving the dynamism. The previous behavior is that we baked in True/False in the cond operator. This can be confusing. In this PR, we change it to be specializing into one of the branches when the inputs are constants.

We additionally change the naming of cond operator to default one without overriding its name. This allows better testing on de-serialized graph.

Test Plan:
The predicate in some existing tests is the result of a shape comparison. When no dynamic shape is involved, the predicate is a python bool. To fix them, we either change the predicate to be some data-dependent tensor or change the test to check cond is specialized as one of the branches, 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D59589709](https://our.internmc.facebook.com/intern/diff/D59589709)

[ghstack-poisoned]

ydwu4 commented:
replaced by https://github.com/pytorch/pytorch/pull/130493
